### PR TITLE
Ensure installed stdlib artefacts have correct case

### PR DESCRIPTION
**Pull request by dra27**
Apropos @garrigue's https://github.com/ocaml/ocaml/pull/10285#pullrequestreview-616081639, this amends the `stdlib/Makefile` install target to remove any `stdlib__` files in the installation libdir.

This fixes the obscure, but not impossible, situation where you install a new compiler over the top of an old one - upgrades via opam or any reputable OS package manager (hopefully, even the disreputable ones!) will have removed the old compiler first.

### CODE REVIEW

**Comment by gasche** [comment]
I'm not convinced by the idea of removing things at the installation directory. Copying new files there makes sense, but I would rather avoid removing anything that is not strictly within the build directory. I like this guarantee better than making things marginally better for people who overwrite an old install directory with a new installation on case-insensitive systems.

**Comment by gasche** [comment]
(Note: if we really want to go the effort, reading what's in the installation directory and showing a warning when it looks problematic would be okay with me.)

**Comment by dra27** [comment]
On macOS and Windows, we're not actually "removing" anything, because we're about to install identical files over the top. However, if we don't remove those existing files first, the newly installed files will have the wrong case and the "successfully" installed compiler will not work.

On Linux (or with a case-sensitive file system) we're removing files which definitely should no longer be there, from "our" libdir.

**Comment by gasche** [comment]
Well let's say that I don't like the idea of an install script calling `rm` on a path outside the build directory.

(Note that we also support things like `$DESTDIR` that are meant to let people install things in a temporary place, to be copied in the correct place later; in general my intuition is that install scripts should act as a big copy of build-directory files into installation directories, and nothing else.)

**Comment by nojb** [comment]
Just as a data point, we already did this at least twice in the past in response to user bug reports #1724 and #1728

**Comment by dra27** [comment]
Thanks, @nojb - I had a memory that we've done this before!

**Comment by dra27** [comment]
An alternative pushed: fail the install/installopt if the stale artefacts are found. We don't offer an `uninstall` target, so I think it's not unreasonable that users should expect that a successful `make install` should have transferred a working compiler from a build tree to be a working compiler in the prefix.

**Comment by gasche** [comment]
I didn't mean to be blocking this PR, just expressing my personal preference (which isn't really changed by the fact that we already chose to remove files in the past). I would be tempted to encourage @nojb to make decisions on this particular PR, if not for the fact that I have already thrown a sizeable amount of work in his direction in the last few days.

**Comment by nojb** [comment]
Just to understand: what are the visible consequences *today* if you happen to install the new compiler over the old compiler and you end up with files with both capitalizations, eg `stdlib__string.cmi` and `stdlib__String.cmi` in the installation directory on one of the affected systems ? do you get an error when the compiler tries to load `stdlib__String.cmi` ?

**Comment by dra27** [comment]
On Linux it makes no difference, but on macOS and Windows you don't up with _both_ capitalisations, you end up with the _wrong_ (old) capitalization. It's this effect (I believe the same on macOS, but I lazily haven't fired up my Macbook!):

```
C:\Users\DRA>echo bar>foo
C:\Users\DRA>echo baz>Foo
C:\Users\DRA>dir foo
19/04/2021  10:41                 5 foo
```

**Comment by nojb** [comment]
Thanks. I'm in favor of merging the present PR: this kind of problem bites a very small number of users, but when it happens it is puzzling and tricky to figure out what is going on.

I think the current version (which doesn't delete anything, just prints a message and fails) is good enough: it will clearly detect the problem but does not delete files from the installation directory which may be somewhat unorthodox.

File: stdlib/Makefile
Comment (nojb) [inline stdlib/Makefile:65]:
```suggestion
    echo "Please rm $(INSTALL_LIBDIR)/stdlib__*.cmi and re-run make install"; \
```
?

**Review by nojb — APPROVED** [review]
LGTM (modulo a tiny question)

File: stdlib/Makefile
Comment (dra27) [reply stdlib/Makefile:65]: Well noticed 🙂 My theory here was that it would be annoying to be told at this stage to delete the .cmi files only to be told when you rerun `make install` that you should now also delete the .cmx files. The other message can only be reached (officially) if the `.cmi` files had already been deleted.

Comment (nojb) [reply stdlib/Makefile:65]: Makes sense, thanks.

Comment (Et7f3) [reply stdlib/Makefile:65]: If cmt and cmti are already present can this cause issue ? The variable stale only check for cmi.

Comment (dra27) [reply stdlib/Makefile:65]: You'd need to have installed 4.07-4.12, manually removed only the .cmi files and then also be in the relatively unusual position of wanting to install 4.13+ over the top of that manually-broken old installation 🙂

Source artefacts are installed in the `install::` rule following this one - there could be the same check there, but it feels somewhat overkill to add it?

**Pull request merged**
